File: Orphanage/dotnetapp/Services/DonationService.cs
using dotnetapp.Data;
using dotnetapp.Models;
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace dotnetapp.Services
{
    public class DonationService
    {
        private readonly ApplicationDbContext _context;

        public DonationService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<IEnumerable<Donation>> GetAllDonations()
        {
            return await _context.Donations.Include(r => r.User).Include(r => r.Orphanage).ToListAsync();
        }
       public async Task<IEnumerable<Donation>> GetDonationsByUserId(int userId)
        {
            return await _context.Donations
                .Where(c => c.UserId == userId)
                .Include(r => r.Orphanage)
                .ToListAsync();
        }
        public async Task<Donation> GetDonationById(int donationId)
        {
            return await _context.Donations.FirstOrDefaultAsync(d => d.DonationId == donationId);
        }

        public async Task<bool> AddDonation(Donation donation)
        {
            _context.Donations.Add(donation);
            await _context.SaveChangesAsync();
            return true;
        }
            public async Task<IEnumerable<Donation>> GetDonationsByOrphanageId(int orphanageId)
        {
            return await _context.Donations
                .Where(d => d.OrphanageId == orphanageId)
                .ToListAsync();
        }

        public async Task<bool> UpdateDonation(int donationId, Donation donation)
        {
            var existingDonation = await _context.Donations.FirstOrDefaultAsync(d => d.DonationId == donationId);

            if (existingDonation == null)
                return false;

            _context.Entry(existingDonation).CurrentValues.SetValues(donation);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> DeleteDonation(int donationId)
        {
            var donation = await _context.Donations.FirstOrDefaultAsync(d => d.DonationId == donationId);
            if (donation == null)
                return false;

            _context.Donations.Remove(donation);
            await _context.SaveChangesAsync();
            return true;
        }
    }
}

File: Orphanage/angularapp/src/app/services/donation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { apiUrl } from 'src/apiconfig';
import { Donation } from '../models/donation.model';

@Injectable({
  providedIn: 'root'
})
export class DonationService {
  public apiUrl = apiUrl; // Update with your API URL

  constructor(private http: HttpClient) { }

  getAllDonations(): Observable<Donation[]> {
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.get<Donation[]>(`${this.apiUrl}/api/donations`, { headers });
  }

  getDonationById(donationId: number): Observable<Donation> {
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.get<Donation>(`${this.apiUrl}/api/donations/${donationId}`, { headers });
  }

  addDonation(donation: Donation): Observable<Donation> {
    console.log('Donation:', donation);
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.post<Donation>(`${this.apiUrl}/api/donations`, donation, { headers });
  }

  updateDonation(donationId: number, donation: Donation): Observable<Donation> {
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.put<Donation>(`${this.apiUrl}/api/donations/${donationId}`, donation, { headers });
  }

  deleteDonation(donationId: number): Observable<void> {
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.delete<void>(`${this.apiUrl}/api/donations/${donationId}`, { headers });
  }

  getDonationsByUserId(userId: number): Observable<Donation[]> {
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.get<Donation[]>(`${this.apiUrl}/api/donations/user/${userId}`, { headers });
  }

  getDonationsByOrphanageId(orphanageId: number): Observable<Donation[]> {
    const headers = new HttpHeaders({
      'Authorization': 'Bearer ' + localStorage.getItem('token')
    });
    return this.http.get<Donation[]>(`${this.apiUrl}/api/donations/orphanage/${orphanageId}`, { headers });
  }
}
